### PR TITLE
Add username_key as an LTI 1.1 Authenticator configurable

### LINKED ISSUE #30: Use usernames instead of user IDs
I'm currently using LTI authentication with Moodle. By default, the user IDs are used to create the JHub users. Is it possible to use the Moodle usernames (or first/last names) instead?

Help is appreciated. Thanks!

### PR DESCRIPTION
**Pull request by jgwerner**
- Adds the `LTI11Authenticator.username_key` to set the username returned by the authenticator to something other than `custom_canvas_user_id` or `user_id` if specified
- Adds basic tests for `LTI11AuthenticateHandler` class
- Updates Readme

Continuation of #14

Closes #30 

cc/ @BenGig

### CODE REVIEW

File: ltiauthenticator/lti11/auth.py
Comment (yuvipanda) [inline ltiauthenticator/lti11/auth.py:52]: Let's mark the canvas_custom_user_id default as deprecated? We no longer need it to be hard coded since people who want this behavior can set `username_key` instead.

File: ltiauthenticator/lti11/auth.py
Comment (yuvipanda) [inline ltiauthenticator/lti11/auth.py:100]:
```suggestion
```

Not actually needed, since we never want to return "" as a username. If `username` isn't defined in all the arms of the `if` statement below, we'll get a clean reference error than dig through to find out why username is `""`

**Review by yuvipanda — CHANGES_REQUESTED** [review]
Awesome! Couple minor things, happy to merge afterwards

**Comment by yuvipanda** [comment]
What do you think of making a release after this gets merged?

**Comment by jgwerner** [comment]
@yuvipanda creating a new release sounds good to me! I'll update this PR based on your comments, thanks for the review.

File: ltiauthenticator/lti11/auth.py
Comment (jgwerner) [reply ltiauthenticator/lti11/auth.py:100]: Of course! Thanks for pointing that out.

File: ltiauthenticator/lti11/auth.py
Comment (jgwerner) [reply ltiauthenticator/lti11/auth.py:52]: Glad you mentioned this. Agree and will make the change. We could set the username_key to default to user_id since it's the best common denominator (including for private launches when things are opaque). I also thought of adding a list of Enums to ensure the username_key is valid for LTI 1.1 but would rather do that in a separate PR.

Comment (yuvipanda) [reply ltiauthenticator/lti11/auth.py:52]: Yeah, I agree re: setting default to `user_id`. Let's mark the `canvas_custom_user_id` as deprecated, emit a warning if the user is relying on that, and then change the default on next release?

Comment (yuvipanda) [reply ltiauthenticator/lti11/auth.py:52]: Mostly since end users will see this breaking change as 'my files are all gone! PANIC!' so want to make sure people have at least some warning

File: README.md
Comment (consideRatio) [inline README.md:58]: Wieeee I really appreciate the level on which you communicate here. It is easy to make a too generous assumption about the users knowledge and not introducing the user to when and why something makes sense to configure, but I think this section does an excellent job.

:heart:

**Review by consideRatio — APPROVED** [review]
@jgwerner wow amazing work!!! I appreciate how thorough you are!

Excellent documentation, excellent testing! :heart: :tada:

Assuming @yuvipanda comments are addressed, this LGTM so I'll press approve at this point!

**Comment by jgwerner** [comment]
@yuvipanda here I updated the PR with the following logic:

- Log a deprecation warning if the username_key = "custom_canvas_user_id"
- Check if the `username_key` is in the launch request
- If the `username_key` is not in the launch request raise an HTTP error
- If the `username` is empty or None, fallback to the `user_id` value
- If the `username` is still empty raise an HTTP error

There were also some duplicate validator tests in the authenticator file so removed them. Perhaps we should have refactored tests in another PR, but wanted to make sure the flow above had passing tests without duplicates.

File: ltiauthenticator/lti11/auth.py
Comment (yuvipanda) [inline ltiauthenticator/lti11/auth.py:54]:
```suggestion
        Current default behavior:
        
        To preserve legacy behavior, if custom_canvas_user_id is present in the LTI
        request, it is used as the username. If not, user_id is used. In the future,
        the default will be just user_id - if you want to use custom_canvas_user_id,
        you must explicitly set username_key to custom_canvas_user_id.
        
```

File: ltiauthenticator/lti11/auth.py
Comment (yuvipanda) [inline ltiauthenticator/lti11/auth.py:85]:
```suggestion
                dedent("""The default username_key 'custom_canvas_user_id' will be replaced by 'user_id' in a future release.
                Set c.LTIAuthenticator.username_key to `custom_canvas_user_id` to preserve current behavior
                """)
```

Might have to `from textwrap import dedent`

**Comment by yuvipanda** [comment]
@jgwerner thank you for your patience :) This lgtm for the most part - I left some documentation change suggestions, what do you think? Happy to merge this after.

**Comment by jgwerner** [comment]
@yuvipanda I update the PR based on your suggestions, thanks for the documentation tips 👍

**Comment by yuvipanda** [comment]
Thank you, @jgwerner!

**Pull request merged**
